src/lib.rs

```diff
@@ -1190,20 +1209,80 @@ pub enum MapMode {
     Write,
 }
 
+fn range_to_offset_size<S: RangeBounds<BufferAddress>>(bounds: S) -> (BufferAddress, BufferSize) {
+    let offset = match bounds.start_bound() {
+        Bound::Included(&bound) => bound,
+        Bound::Excluded(&bound) => bound + 1,
+        Bound::Unbounded => 0,
+    };
+    let size = match bounds.end_bound() {
+        Bound::Included(&bound) => BufferSize(bound + 1 - offset),
+        Bound::Excluded(&bound) => BufferSize(bound - offset),
+        Bound::Unbounded => BufferSize::WHOLE,
+    };
+
+    (offset, size)
+}
+
+pub struct BufferView<'a> {
+    buffer: &'a Buffer,
+    data: &'a [u8],
+    offset: BufferAddress,
+    size: BufferSize,
+}
+
+pub struct BufferViewMut<'a> {
+    buffer: &'a Buffer,
+    data: &'a mut [u8],
+    offset: BufferAddress,
+    size: BufferSize,
+}
+
+impl std::ops::Deref for BufferView<'_> {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        self.data
+    }
+}
+
+impl std::ops::Deref for BufferViewMut<'_> {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        self.data
```

Comment:
Can we also have a flag in `BufferViewMut`, something like `readable: bool`?

This would allow us to add the following check on `deref()`:
```rust
assert!(self.readable, "Attempting to read a write-only mapping for buffer {:?}", self.slice.buffer.id);
```
Reading a write-only mapping may have unintended consequences: the implementation may have a write-combined memory type on native, reading from which is horribly slow. We don't want these surprises.

Reply:
Good idea. Didn't realize there was write-only memory.

Reply:
Strictly speaking, users can still read from `&mut []` that they get from `DerefMut`, but that one is much harder to prevent, so it's ok.

Reply:
There may be some way of having a write-only slice with a new-type.